code/modules/projectiles/modular/laser_base.dm

```diff
@@ -0,0 +1,215 @@
+	. = ..(user, 1)
+	if(.)
+		if(condition > 0)
+			user << "<span class='warning'>\The [src] appears damaged.</span>"
+
+/obj/item/laser_components/capacitor/proc/small_fail(var/obj/item/weapon/gun/energy/laser/prototype/prototype)
+	return
+
+/obj/item/laser_components/capacitor/proc/medium_fail(var/obj/item/weapon/gun/energy/laser/prototype/prototype)
+	return
+
+/obj/item/laser_components/capacitor/proc/critical_fail(var/obj/item/weapon/gun/energy/laser/prototype/prototype)
+	qdel(src)
+	return
+
+/obj/item/laser_components/focusing_lens
+	name = "focusing lens"
+	desc = "A basic laser weapon focusing lens."
+	icon_state = "lens"
+	var/dispersion = 0
+	reliability = 25
+	repair_item = /obj/item/stack/nanopaste
+
+/obj/item/laser_components/focusing_lens/negadegrade(var/obj/item/weapon/D)
+	var/obj/item/stack/nanopaste/N = D
+	if(N.amount)
+		N.amount -= 5
+		condition -= reliability/5
+		if(condition < 0)
+			condition = 0
+		return 1
+	return 0
+
+/obj/item/laser_components/focusing_lens/examine(mob/user)
+	. = ..(user, 1)
+	if(.)
+		if(condition > 0)
+			user << "<span class='warning'>\The [src] appears damaged.</span>"
+
+/obj/item/laser_components/modulator
+	name = "laser modulator"
+	desc = "A modification that modulates the beam into a standard laser beam."
+	icon_state = "laser"
+	var/obj/item/projectile/beam/projectile = /obj/item/projectile/beam
+
+/obj/item/laser_components/modulator/degrade()
+	return
+
+/obj/item/device/laser_assembly
+	name = "laser assembly (small)"
+	desc = "A case for shoving things into. Hopefully they work."
+	w_class = 2
+	icon = 'icons/obj/modular_laser.dmi'
+	icon_state = "small"
+	var/stage = 1
+	var/size = CHASSIS_SMALL
+	var/modifier_cap = 1
+
+	var/list/gun_mods = list()
+	var/obj/item/laser_components/capacitor/capacitor
+	var/obj/item/laser_components/focusing_lens/focusing_lens
+	var/obj/item/laser_components/modulator/modulator
+
+/obj/item/device/laser_assembly/Initialize()
+	..()
+	update_icon()
+
+/obj/item/device/laser_assembly/attackby(var/obj/item/weapon/D as obj, var/mob/user as mob)
+	var/obj/item/laser_components/A = D
+	if(!istype(A))
+		return ..()
+	if(ismodifier(A) && gun_mods.len < modifier_cap)
+		gun_mods += A
+		user.drop_item()
+		A.forceMove(src)
+	else if(islasercapacitor(A) && stage == 1)
+		capacitor = A
+		user.drop_item()
+		A.forceMove(src)
+		stage = 2
+	else if(isfocusinglens(A) && stage == 2)
+		focusing_lens = A
+		user.drop_item()
+		A.forceMove(src)
+		stage = 3
+	else if(ismodulator(A) && stage == 3)
+		modulator = A
+		user.drop_item()
+		A.forceMove(src)
+	else
+		return ..()
+	user << "<span class='notice'>You insert the [A] into the assembly.</span>"
+	update_icon()
+	check_completion()
+
+/obj/item/device/laser_assembly/update_icon()
+	..()
+	underlays.Cut()
+	icon_state = "[initial(icon_state)]_[stage]"
+	if(gun_mods.len)
+		for(var/obj/item/laser_components/mod in gun_mods)
+			if(mod.gun_overlay)
+				underlays += mod.gun_overlay
```

Comment:
This should probably use the overlay underlay shit? Unless not, check with Lohikar.

Reply:
I asked Lohikar and he said that only overlays had an overlay() proc because apparently I'm the only madman that ever uses underlays so queuing them is not expensive.